lib/SlackEventHandler.js

```diff
@@ -0,0 +1,222 @@
+                    } else if (e instanceof UnknownEvent) {
+                        endTimer({outcome: "dropped"});
+                    } else {
+                        endTimer({outcome: "fail"});
+                    }
+
+                    console.log("Failed: ", e);
+                }
+            );
+
+        response.writeHead(200, {"Content-Type": "application/json"});
+        response.end();
+    } catch (e) {
+        console.log("Oops - SlackEventHandler failed:", e);
+
+        // return 200 so slack doesn't keep sending the event
+        response.writeHead(200, {"Content-Type": "text/plain"});
+        response.end();
+    }
+};
+
+/**
+ * Attempts to handle the `team_domain_change` event.
+ *
+ * @param {Object} params The event request emitted.
+ * @param {Object} params.team_id The slack team_id for the event.
+ * @param {string} params.event.domain The new team domain.
+ */
+SlackEventHandler.prototype.handleDomainChangeEvent = function (params) {
+    this._main.getRoomsBySlackTeamId(params.team_id).forEach(room => {
+        room.updateSlackTeamDomain(params.event.domain);
+        if (room.isDirty()) {
+            this._main.putRoomToStore(room);
+        }
+    });
+    return Promise.resolve();
+};
+
+/**
+ * Attempts to handle the `channel_rename` event.
+ *
+ * @param {Object} params The event request emitted.
+ * @param {string} params.event.id The slack channel id
+ * @param {string} params.event.name The new name
+ */
+SlackEventHandler.prototype.handleChannelRenameEvent = function (params) {
+    //TODO test me. and do we even need this? doesn't appear to be used anymore
+    var room = this._main.getRoomBySlackChannelId(params.event.channel.id);
+    if (!room) throw new UnknownChannel(params.event.channel.id);
+
+    var channel_name = room.getSlackTeamDomain() + ".#" + params.name;
+    room.updateSlackChannelName(channel_name);
+    if (room.isDirty()) {
+        this._main.putRoomToStore(room);
+    }
+    return Promise.resolve();
+};
+
+/**
+ * Attempts to handle the `message` event.
+ *
+ * Sends a message to Matrix if it understands enough of the message to do so.
+ * Attempts to make the message as native-matrix feeling as it can.
+ *
+ * @param {Object} params The event request emitted.
+ * @param {string} params.event.user Slack user ID of user sending the message.
+ // * @param {string} params.user_name Slack user name of the user sending the message.
+ * @param {?string} params.event.text Text contents of the message, if a text message.
+ * @param {string} params.event.channel The slack channel id
+ * @param {string} params.event.ts The unique (per-channel) timestamp
+ */
+SlackEventHandler.prototype.handleMessageEvent = function (params) {
+    var room = this._main.getRoomBySlackChannelId(params.event.channel);
+    if (!room) throw new UnknownChannel(params.event.channel);
+
+    if (params.event.subtype === 'bot_message' && params.event.bot_id === room.getSlackBotId()) {
```

Comment:
If there is not `slack_bot_token`, `lookupAndSetUserInfo` will never set `slack_bot_id`, so this condition is always true, and I'm afraid we have a loop.

The condition should be something like:
```Javascript
params.event.subtype === 'bot_message' && (!room.getSlackBotToken || params.event.bot_id === room.getSlackBotId())
```

Reply:
I will let you answer, but if I understand your code we got here only if we are using the Event API, so `slack_bot_id` _is_ defined.

Reply:
yes, the `SlackEventHandler` should only be used if `slack_bot_id` is set. I added your check anyways, as it should save some loops when incorrectly setup